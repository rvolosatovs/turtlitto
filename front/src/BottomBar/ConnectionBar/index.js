import React, { Fragment } from "react";
import styled, { css } from "styled-components";
import PropTypes from "prop-types";
import connectionTypes from "../connectionTypes";
import FontAwesomeIcon from "@fortawesome/react-fontawesome";
import faSpinner from "@fortawesome/fontawesome-free-solid/faSpinner";

const Bar = styled.div`
  ${props => props.background};
  padding: 0.6rem 3rem 0.6rem 0.6rem;
  display: flex;
  align-items: center;
  justify-content: center;
  color: white;
  position: relative;
`;

const Message = styled.p`
  text-transform: uppercase;
  margin: 0;
  display: block;
  font-size: 1.4rem;
`;

const IconWrapper = styled.div`
  padding-left: 2rem;
  position: absolute;
  top: 50%;
  right: 1rem;
  transform: translateY(-50%);
`;

const getBackground = type => {
  switch (type) {
    case connectionTypes.CONNECTING:
      return css`
        background: ${props => props.theme.notificationWarning};
      `;
    case connectionTypes.CONNECTED:
      return css`
        background: ${props => props.theme.notificationSuccess};
      `;
    case connectionTypes.DISCONNECTED:
      return css`
        background: ${props => props.theme.notificationError};
      `;
    default:
      throw new Error("Unknown connection status type");
  }
};

const getContent = type => {
  switch (type) {
    case connectionTypes.CONNECTING:
      return (
        <Fragment>
          <Message>connecting...</Message>
          <IconWrapper>
            <FontAwesomeIcon size="2x" spin icon={faSpinner} />
          </IconWrapper>
        </Fragment>
      );
    case connectionTypes.CONNECTED:
      return <Message>connected</Message>;
    case connectionTypes.DISCONNECTED:
      return <Message>diconnected</Message>;
    default:
      throw new Error("Unknown connection status type");
  }
};

const ConnectionBar = props => {
  const background = getBackground(props.connectionStatus);
  const content = getContent(props.connectionStatus);
  return (
    <Bar className={props.className} background={background}>
      {content}
    </Bar>
  );
};

ConnectionBar.propTypes = {
  connectionStatus: PropTypes.oneOf(Object.values(connectionTypes)).isRequired
};

export default ConnectionBar;
