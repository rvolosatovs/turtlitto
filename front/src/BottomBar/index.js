import styled from "styled-components";
import FontAwesomeIcon from "@fortawesome/react-fontawesome";
import faPlay from "@fortawesome/fontawesome-free-solid/faPlay";
import faStop from "@fortawesome/fontawesome-free-solid/faStop";
import faCog from "@fortawesome/fontawesome-free-solid/faCog";
import faThLarge from "@fortawesome/fontawesome-free-solid/faThLarge";
import React from "react";
import PropTypes from "prop-types";
import pageTypes from "./pageTypes";
import connectionTypes from "./connectionTypes";
import sendToServer from "../sendToServer";
import { media } from "../media";
import ConnectionBar from "./ConnectionBar";

/**
 * All of the props depend on the implementation in App.js. App.js should include WebSockets,
 * a state variable which has the values "settings" and "refbox",
 * and the implementation of the functions passed to BottomBar.
 * Author: T.T.P. Franken
 * Author: B. Afonins
 * Author: G.M. van der Sanden
 *
 * Props:
 * - changeActivePage: function to change the active page
 * - activePage: a string indicating the current active page
 * - isConnected: a boolean indicating whether the client is connected to the TRC
 */

const BottomBar = props => {
  const { changeActivePage, activePage, connectionStatus } = props;
  const isSettingsPage = activePage === pageTypes.SETTINGS;

  return (
    <Bar className={props.className}>
      <ConnectionBar connectionStatus={connectionStatus} />
      <ButtonsWrapper>
        <ButtonColumn>
          <StartButton
            id="bottom-bar__start-button"
            onClick={() => {
              sendToServer("start", "command");
            }}
            enabled
          >
            <FontAwesomeIcon icon={faPlay} />
          </StartButton>
          <ChangePageButton
            id="bottom-bar__change-page-button"
            onClick={() =>
              changeActivePage(
                isSettingsPage ? pageTypes.REFBOX : pageTypes.SETTINGS
              )
            }
            enabled
          >
            {isSettingsPage ? (
              <FontAwesomeIcon icon={faThLarge} />
            ) : (
              <FontAwesomeIcon icon={faCog} />
            )}
          </ChangePageButton>
        </ButtonColumn>
        <StopButton
          id="bottom-bar__stop-button"
          onClick={() => {
            sendToServer("stop", "command");
          }}
          enabled
        >
          <FontAwesomeIcon icon={faStop} color="red" />
        </StopButton>
      </ButtonsWrapper>
    </Bar>
  );
};

BottomBar.propTypes = {
  changeActivePage: PropTypes.func.isRequired,
  activePage: PropTypes.oneOf(Object.values(pageTypes)),
  connectionStatus: PropTypes.oneOf(Object.values(connectionTypes)).isRequired
};

const ButtonsWrapper = styled.div`
  display: flex;
  justify-content: space-between;
  margin: 0rem;
  flex: 1;
`;

const ButtonColumn = styled.div`
  width: 50%;
  height: 100%;
  justify-content: space-between;
  margin: 0rem;
  flex: 1;
`;

const Button = styled.button`
  background: ${props => props.theme.bottomBarButton};
  width: 100%;
  font-size: 2rem;
  flex: 1;

  &:active {
    background: ${props => props.theme.bottomBarButtonActive};
  }
`;

const StopButton = styled(Button)`
  font-size: 5rem;
`;

const StartButton = styled(Button)`
  height: 50%;

  ${media.md`
    height: 100%;
    font-size: 5rem;
  `};
`;

const ChangePageButton = styled(Button)`
  height: 50%;
  display: inline-block;

  ${media.md`
    display: none;
  `};
`;

const Bar = styled.div`
  position: sticky;
  background-color: white;
  top: 100%;
  width: 100%;
  height: 20%;
  display: flex;
  flex-direction: column;
  margin: 0px;
`;

export default BottomBar;
