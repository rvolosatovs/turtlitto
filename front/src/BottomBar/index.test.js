import React from "react";
import BottomBar from ".";
import { shallow } from "enzyme";
import sinon from "sinon";
import connectionTypes from "./connectionTypes";
import pageTypes from "./pageTypes";

describe("BottomBar", () => {
  describe("is in the refbox mode", () => {
    it("should match snapshot", () => {
      const wrapper = shallow(
        <BottomBar
          changeActivePage={() => {}}
          activePage={pageTypes.REFBOX}
          connectionStatus={connectionTypes.CONNECTED}
        />
      );

      expect(wrapper).toMatchSnapshot();
    });

    describe("the user clicks on the settings button", () => {
      it("should pass `settings` to the `changeActivePage` function", () => {
        const changeActivePageSpy = sinon.spy();
        const wrapper = shallow(
          <BottomBar
            changeActivePage={changeActivePageSpy}
            activePage={pageTypes.REFBOX}
            connectionStatus={connectionTypes.CONNECTED}
          />
        );

        wrapper.find("#bottom-bar__change-page-button").simulate("click");

        expect(changeActivePageSpy.calledOnce).toBe(true);
        expect(changeActivePageSpy.calledWithExactly("settings")).toBe(true);
      });
    });
  });

  describe("is in the settings mode", () => {
    it("should match snapshot", () => {
      const wrapper = shallow(
        <BottomBar
          changeActivePage={() => {}}
          activePage={pageTypes.SETTINGS}
          connectionStatus={connectionTypes.CONNECTED}
        />
      );

      expect(wrapper).toMatchSnapshot();
    });

    describe("the user clicks on the refbox button", () => {
      it("should pass `refbox` to the `changeActivePage` function", () => {
        const changeActivePageSpy = sinon.spy();
        const wrapper = shallow(
          <BottomBar
            changeActivePage={changeActivePageSpy}
            activePage={pageTypes.SETTINGS}
            connectionStatus={connectionTypes.CONNECTED}
          />
        );

        wrapper.find("#bottom-bar__change-page-button").simulate("click");

        expect(changeActivePageSpy.calledOnce).toBe(true);
        expect(changeActivePageSpy.calledWithExactly("refbox")).toBe(true);
      });
    });
  });
});

describe("When clicked,", () => {
  const realFetch = global.fetch;
  let wrapper = null;
  beforeEach(() => {
    wrapper = shallow(
      <BottomBar
        changeActivePage={() => {}}
        activePage={pageTypes.REFBOX}
        connectionStatus={connectionTypes.CONNECTED}
      />
    );
    const l = window.location;
    global.fetch = jest.fn().mockImplementation((url, params) => {
      expect(url).toBe(`${l.protocol}//${l.host}/api/v1/command`);
      expect(params).toMatchSnapshot();
      return Promise.resolve({ ok: true });
    });
  });

  afterEach(() => {
    global.fetch = realFetch;
  });
  it("the start button should pass `start` to the 'sendToServer' function", () => {
    wrapper.find("#bottom-bar__start-button").simulate("click");
  });
  it("the stop button should pass `stop` to the 'sendToServer' function", () => {
    wrapper.find("#bottom-bar__stop-button").simulate("click");
  });
});
