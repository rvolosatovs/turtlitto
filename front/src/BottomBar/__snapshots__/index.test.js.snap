// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BottomBar is in the refbox mode should match snapshot 1`] = `
<styled.div>
  <ConnectionBar
    connectionStatus="connected"
  />
  <styled.div>
    <styled.div>
      <Styled(styled.button)
        enabled={true}
        id="bottom-bar__start-button"
        onClick={[Function]}
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                448,
                512,
                Array [],
                "f04b",
                "M424.4 214.7L72.4 6.6C43.8-10.3 0 6.1 0 47.9V464c0 37.5 40.7 60.1 72.4 41.3l352-208c31.4-18.5 31.5-64.1 0-82.6z",
              ],
              "iconName": "play",
              "prefix": "fas",
            }
          }
          listItem={false}
          mask={null}
          name=""
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          symbol={false}
          transform={null}
        />
      </Styled(styled.button)>
      <Styled(styled.button)
        enabled={true}
        id="bottom-bar__change-page-button"
        onClick={[Function]}
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                512,
                512,
                Array [],
                "f013",
                "M444.788 291.1l42.616 24.599c4.867 2.809 7.126 8.618 5.459 13.985-11.07 35.642-29.97 67.842-54.689 94.586a12.016 12.016 0 0 1-14.832 2.254l-42.584-24.595a191.577 191.577 0 0 1-60.759 35.13v49.182a12.01 12.01 0 0 1-9.377 11.718c-34.956 7.85-72.499 8.256-109.219.007-5.49-1.233-9.403-6.096-9.403-11.723v-49.184a191.555 191.555 0 0 1-60.759-35.13l-42.584 24.595a12.016 12.016 0 0 1-14.832-2.254c-24.718-26.744-43.619-58.944-54.689-94.586-1.667-5.366.592-11.175 5.459-13.985L67.212 291.1a193.48 193.48 0 0 1 0-70.199l-42.616-24.599c-4.867-2.809-7.126-8.618-5.459-13.985 11.07-35.642 29.97-67.842 54.689-94.586a12.016 12.016 0 0 1 14.832-2.254l42.584 24.595a191.577 191.577 0 0 1 60.759-35.13V25.759a12.01 12.01 0 0 1 9.377-11.718c34.956-7.85 72.499-8.256 109.219-.007 5.49 1.233 9.403 6.096 9.403 11.723v49.184a191.555 191.555 0 0 1 60.759 35.13l42.584-24.595a12.016 12.016 0 0 1 14.832 2.254c24.718 26.744 43.619 58.944 54.689 94.586 1.667 5.366-.592 11.175-5.459 13.985L444.788 220.9a193.485 193.485 0 0 1 0 70.2zM336 256c0-44.112-35.888-80-80-80s-80 35.888-80 80 35.888 80 80 80 80-35.888 80-80z",
              ],
              "iconName": "cog",
              "prefix": "fas",
            }
          }
          listItem={false}
          mask={null}
          name=""
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          symbol={false}
          transform={null}
        />
      </Styled(styled.button)>
    </styled.div>
    <Styled(styled.button)
      enabled={true}
      id="bottom-bar__stop-button"
      onClick={[Function]}
    >
      <FontAwesomeIcon
        border={false}
        className=""
        color="red"
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              448,
              512,
              Array [],
              "f04d",
              "M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48z",
            ],
            "iconName": "stop",
            "prefix": "fas",
          }
        }
        listItem={false}
        mask={null}
        name=""
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        symbol={false}
        transform={null}
      />
    </Styled(styled.button)>
  </styled.div>
</styled.div>
`;

exports[`BottomBar is in the settings mode should match snapshot 1`] = `
<styled.div>
  <ConnectionBar
    connectionStatus="connected"
  />
  <styled.div>
    <styled.div>
      <Styled(styled.button)
        enabled={true}
        id="bottom-bar__start-button"
        onClick={[Function]}
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                448,
                512,
                Array [],
                "f04b",
                "M424.4 214.7L72.4 6.6C43.8-10.3 0 6.1 0 47.9V464c0 37.5 40.7 60.1 72.4 41.3l352-208c31.4-18.5 31.5-64.1 0-82.6z",
              ],
              "iconName": "play",
              "prefix": "fas",
            }
          }
          listItem={false}
          mask={null}
          name=""
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          symbol={false}
          transform={null}
        />
      </Styled(styled.button)>
      <Styled(styled.button)
        enabled={true}
        id="bottom-bar__change-page-button"
        onClick={[Function]}
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                512,
                512,
                Array [],
                "f009",
                "M296 32h192c13.255 0 24 10.745 24 24v160c0 13.255-10.745 24-24 24H296c-13.255 0-24-10.745-24-24V56c0-13.255 10.745-24 24-24zm-80 0H24C10.745 32 0 42.745 0 56v160c0 13.255 10.745 24 24 24h192c13.255 0 24-10.745 24-24V56c0-13.255-10.745-24-24-24zM0 296v160c0 13.255 10.745 24 24 24h192c13.255 0 24-10.745 24-24V296c0-13.255-10.745-24-24-24H24c-13.255 0-24 10.745-24 24zm296 184h192c13.255 0 24-10.745 24-24V296c0-13.255-10.745-24-24-24H296c-13.255 0-24 10.745-24 24v160c0 13.255 10.745 24 24 24z",
              ],
              "iconName": "th-large",
              "prefix": "fas",
            }
          }
          listItem={false}
          mask={null}
          name=""
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          symbol={false}
          transform={null}
        />
      </Styled(styled.button)>
    </styled.div>
    <Styled(styled.button)
      enabled={true}
      id="bottom-bar__stop-button"
      onClick={[Function]}
    >
      <FontAwesomeIcon
        border={false}
        className=""
        color="red"
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              448,
              512,
              Array [],
              "f04d",
              "M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48z",
            ],
            "iconName": "stop",
            "prefix": "fas",
          }
        }
        listItem={false}
        mask={null}
        name=""
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        symbol={false}
        transform={null}
      />
    </Styled(styled.button)>
  </styled.div>
</styled.div>
`;

exports[`When clicked, the start button should pass \`start\` to the 'sendToServer' function 1`] = `
Object {
  "body": "\\"start\\"",
  "headers": Object {},
  "method": "PUT",
}
`;

exports[`When clicked, the stop button should pass \`stop\` to the 'sendToServer' function 1`] = `
Object {
  "body": "\\"stop\\"",
  "headers": Object {},
  "method": "PUT",
}
`;
